common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcher.kt

```diff
@@ -0,0 +1,103 @@
+) {
+
+    /**
+     * Queries the list of purchased products for a given user from the billing system,
+     * and maps them to `PurchasedProduct` objects.
+     *
+     * @param appUserID The ID of the user for which to retrieve the purchased products.
+     * @param onSuccess A callback function that will be called with the list of `PurchasedProduct`
+     * objects upon a successful query.
+     * @param onError A callback function that will be called with a `PurchasesError` object
+     * upon a failed query.
+     */
+    fun queryPurchasedProducts(
+        appUserID: String,
+        onSuccess: (List<PurchasedProduct>) -> Unit,
+        onError: (PurchasesError) -> Unit
+    ) {
+        val productEntitlementMapping = deviceCache.getProductEntitlementMapping()
+
+        billing.queryAllPurchases(
+            appUserID,
+            onReceivePurchaseHistory = { allPurchases ->
+                billing.queryPurchases(
+                    appUserID,
+                    onSuccess = { activePurchasesByHashedToken ->
+                        val purchasedProductIdentifiers = allPurchases.map { it.skus[0] }.toSet()
+                        val activeProducts = activePurchasesByHashedToken.values.map { it.skus[0] }.toSet()
+                        val purchasedProducts = purchasedProductIdentifiers.map { productIdentifier ->
+                            createPurchasedProduct(
+                                allPurchases,
+                                productIdentifier,
+                                activeProducts,
+                                productEntitlementMapping
+                            )
+                        }
+                        onSuccess(purchasedProducts)
+                    },
+                    onError
+                )
+            },
+            onError
+        )
+    }
+
+    private fun createPurchasedProduct(
+        allPurchases: List<StoreTransaction>,
+        productIdentifier: String,
+        activeProducts: Set<String>,
+        productEntitlementMapping: ProductEntitlementMapping?
+    ): PurchasedProduct {
+        val purchaseAssociatedToProduct = allPurchases.first { it.skus[0] == productIdentifier }
+        val isActive = activeProducts.contains(productIdentifier)
+        val expirationDate = getExpirationDate(isActive, purchaseAssociatedToProduct)
+        return PurchasedProduct(
+            productIdentifier,
+            purchaseAssociatedToProduct,
+            isActive,
+            productEntitlementMapping?.toMap()?.get(productIdentifier) ?: emptyList(),
+            expirationDate
+        )
+    }
+
+    private fun getExpirationDate(
+        isActive: Boolean,
+        purchaseAssociatedToProduct: StoreTransaction
+    ): Date? {
+        return if (purchaseAssociatedToProduct.type == ProductType.SUBS) {
+            if (isActive) {
+                Date(dateProvider.now.time + TimeUnit.DAYS.toMillis(1))
```

Comment:
I did that before and changed it to +1 day because this is what the backend does in the load shedder because it's probably enough. The subscription renewal period is not even that accurate, since the purchase could be a trial, which has a shorter period

Reply:
Yeah, I think this is probably enough and simplifies the logic.